compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,286 @@
+                    )) => {
+                        let ty = lhs.ty(local_decls, tcx).ty;
+
+                        let source_info = st.source_info;
+                        let span = source_info.span;
+
+                        let (num_variants, sizes) =
+                            if let Some(cand) = Self::candidate(tcx, ty, body_did) {
+                                cand
+                            } else {
+                                return None;
+                            };
+                        let adt_def = ty.ty_adt_def().unwrap();
+
+                        let alloc = if let Some(alloc) = alloc_cache.get(ty) {
+                            alloc
+                        } else {
+                            let data_layout = tcx.data_layout();
+                            let ptr_sized_int = data_layout.ptr_sized_integer();
+                            let target_bytes = ptr_sized_int.size().bytes() as usize;
+                            let mut data = vec![0; target_bytes * num_variants as usize];
+                            let mut curr = 0;
+                            for sz in sizes {
+                                match ptr_sized_int {
+                                    rustc_target::abi::Integer::I32 => {
+                                        let bytes = sz.bytes() as u32;
+                                        let bytes = match data_layout.endian {
+                                            rustc_target::abi::Endian::Little => {
+                                                bytes.to_le_bytes()
+                                            }
+                                            rustc_target::abi::Endian::Big => bytes.to_be_bytes(),
+                                        };
+                                        for b in bytes {
+                                            data[curr] = b;
+                                            curr += 1;
+                                        }
+                                    }
+                                    rustc_target::abi::Integer::I64 => {
+                                        let bytes = sz.bytes();
+                                        let bytes = match data_layout.endian {
+                                            rustc_target::abi::Endian::Little => {
+                                                bytes.to_le_bytes()
+                                            }
+                                            rustc_target::abi::Endian::Big => bytes.to_be_bytes(),
+                                        };
+                                        for b in bytes {
+                                            data[curr] = b;
+                                            curr += 1;
+                                        }
+                                    }
+                                    _ => unreachable!(),
+                                };
+                            }
```

Comment:
These match arms look almost identical; is it possible to do just the part that's different in the match arms to make it easier to read? It also might be a bit clearer to use `sizes.enumerate()` instead of the `curr` counter since it looks like it's only used inside the `for` loop.

Reply:
Certainly it is ugly... but I found it hard to consolidate into something more elegant because the `to_*_bytes` returns either a `[u8; 4]` or `[u8; 8]`, and handling them jointly is a bit annoying. I could write a function out with a generic parameter, but just wanted to get it out for now

Reply:
The implementation found in the interpreter I mentioned earlier quite nicely avoids special casing the pointer width.

It operates on largest supported pointer width. Based on endian encodes value using `to_le_bytes` or `to_be_bytes`, and then extracts the least significant portion of data by subslicing. For little endian it extracts header and for big endian tail portion of the array. The extracted slice has the size corresponding to the size of pointer width.